website/addons/dropbox/model.py

```diff
@@ -2,61 +2,95 @@
 from website.addons.base import StorageAddonBase
+from website.oauth.models import ExternalProvider
 
-from website.addons.dropbox.utils import DropboxNodeLogger
+from website.addons.dropbox import settings
+from website.addons.dropbox.serializer import DropboxSerializer
 
 logger = logging.getLogger(__name__)
 
 
-class DropboxUserSettings(AddonUserSettingsBase):
-    """Stores user-specific dropbox information, including the Oauth access
-    token.
-    """
+class DropboxProvider(ExternalProvider):
 
-    dropbox_id = fields.StringField(required=False)
-    access_token = fields.StringField(required=False)
-    dropbox_info = fields.DictionaryField(required=False)
+    name = 'DropBox'
+    short_name = 'dropbox'
 
-    # TODO(sloria): The `user` param in unnecessary for AddonUserSettings
-    def to_json(self, user=None):
-        """Return a dictionary representation of the user settings.
-        The dictionary keys and values will be available as variables in
-        dropbox_user_settings.mako.
-        """
-        output = super(DropboxUserSettings, self).to_json(self.owner)
-        output['has_auth'] = self.has_auth
-        return output
+    client_id = settings.DROPBOX_KEY
+    client_secret = settings.DROPBOX_SECRET
+
+    # Explicitly override auth_url_base as None -- BoxOAuth2Flow handles this for us
+    auth_url_base = None
+    callback_url = None
+    handle_callback = None
 
     @property
-    def has_auth(self):
-        return bool(self.access_token)
+    def oauth_flow(self):
+        if 'oauth_states' not in session.data:
+            session.data['oauth_states'] = {}
+        if self.short_name not in session.data['oauth_states']:
+            session.data['oauth_states'][self.short_name] = {
+                'state': None
+            }
+        return DropboxOAuth2Flow(
+            self.client_id,
+            self.client_secret,
+            redirect_uri=web_url_for(
+                'oauth_callback',
+                service_name=self.short_name,
+                _absolute=True
+            ),
+            session=session.data['oauth_states'][self.short_name], csrf_token_session_key='state'
+        )
 
-    def delete(self, save=True):
-        self.clear()
-        super(DropboxUserSettings, self).delete(save)
-
-    def clear(self):
-        """Clear settings and deauthorize any associated nodes."""
-        self.dropbox_id = None
-        self.access_token = None
-        for node_settings in self.dropboxnodesettings__authorized:
-            node_settings.deauthorize(Auth(self.owner))
-            node_settings.save()
-        return self
+    @property
+    def auth_url(self):
+        return self.oauth_flow.start('force_reapprove=true')
+
+    # Overrides ExternalProvider
+    def auth_callback(self, user):
+        try:
+            access_token, dropbox_user_id, url_state = self.oauth_flow.finish(request.values)
+        except DropboxOAuth2Flow.NotApprovedException:
```

Comment:
Why does this no longer handle `DropboxOAuth2Flow.BadRequestException`, `DropboxOAuth2Flow.BadStateException`, `DropboxOAuth2Flow.CsrfException`, or `DropboxOAuth2Flow.ProviderException`?